### PR TITLE
Speed up default.qubit by using permutations to apply some gates

### PR DESCRIPTION
**Pull request by trbromley**
**Context:**

This PR uses some array/tensor manipulation tricks such as [roll](https://numpy.org/doc/stable/reference/generated/numpy.roll.html), [transpose](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html) and [slicing](https://stackoverflow.com/questions/42656930/numpy-assignment-like-numpy-take) to speed up the application of some gates:
- PauliX
- PauliY
- PauliZ
- Hadamard
- SWAP
- S
- T
- CNOT
- CZ

For example, `PauliX` can be achieved simply using `np.roll(state, 1, wire_axis)`. Currently in PL, these gates are applied using the `_apply_diagonal_unitary()` and `_apply_unitary_einsum()` methods, which both internally perform a contraction using `einsum`.

**Description of the Change:**

Additional methods have been added to `DefaultQubit` for each of the above gates, e.g., `_apply_x()` for `PauliX`. In the main `apply_operation()` method, we then dispatch to each of the functions using a dictionary that maps from operation name to corresponding method.

**Benefits:**

Increased speed. This can be checked using a benchmark script (see at the bottom). The script makes a trial circuit composed of a single gate that is repeatedly placed on each qubit and then repeated for a chosen depth. The script then times how long multiple runs take to complete. Using this, we can benchmark the new and old approaches in `DefaultQubit` and find the speedup as a function of qubit number:
image
We see a speed up, with different gates performing better than others. `PauliY` and `Hadamard` are the worst performers, possibly because they combine two applications of `PauliX` and `PauliZ`.

Note that the changes in this PR are also compatible with `default.qubit.tf` and `default.qubit.autograd`. The speedup plots are shown below:
image
image

I'm not sure exactly why the `default.qubit.tf` plot looks different. Also, these benchmarks are quite artificial in that we only restrict to the gates that we hope to have sped up - practical circuits may include other gates.

**Possible Drawbacks:**

- Edge cases where the new approach is slower than the old?
- `DefaultQubit` feels a bit busy now with a lot of methods. I couldn't pull out these methods as functions in a separate `ops` file because we need access to, e.g., `self._roll` for the interface-correct function.
- Do we need to test these methods explicitly? We don't for other methods in `DefaultQubit`.
- We don't speed up any gates with trainable parameters such as `Rot`, which are quite common. I don't think there is any fundamental roadblock for doing this though, could be a follow up PR.

**Benchmark script:**
```python
import pennylane as qml
import time
import numpy as np

sped_up_gates = [
    qml.PauliX,
    qml.PauliY,
    qml.PauliZ,
    qml.Hadamard,
    qml.SWAP,
    qml.S,
    qml.T,
    qml.CNOT,
    qml.CZ,
]

layers = 5

def get_time(num_wires, layers, reps, gate):

    gate_num_wires = gate.num_wires
    print("Benchmarking for {} qubits on gate {}".format(num_wires, gate))
    
    dev = qml.device("default.qubit.tf", wires=num_wires)
    
    @qml.qnode(dev, interface="tf")
    def benchmark():
        for l in range(layers):
            for i in range(num_wires):
                wires = [w % num_wires for w in range(i, i + gate_num_wires)]
                gate(wires=wires)
        return qml.expval(qml.PauliZ(0))

    time0 = time.time()

    for r in range(reps):
        benchmark()

    time1 = time.time()

    total_time = time1 - time0
    return total_time

def get_time_over_qubits(range_qubits, layers, reps, gate):
    return [get_time(qubit, layers, reps, gate) for qubit in range_qubits]

def get_time_over_gates(range_qubits, layers, reps, range_gates):
    return np.array([get_time_over_qubits(range_qubits, layers, reps, gate) for gate in range_gates])

results = get_time_over_gates(range(2, 8), layers, 1000, sped_up_gates)
```

### CODE REVIEW

**Comment by co9olguy** [comment]
Won't have a chance to formally code review, but had a quick scan and looks great @trbromley!

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -84,6 +84,8 @@ class QubitDevice(Device):
     _tensordot = staticmethod(np.tensordot)
     _conj = staticmethod(np.conj)
     _imag = staticmethod(np.imag)
+    _roll = staticmethod(np.roll)
+    _stack = staticmethod(np.stack)
```
Comment (josh146) [inline pennylane/_qubit_device.py:88]: Just double checking, but something I noticed in previous PRs is that the three frameworks don't have consistency in how they implement 'stacking'. e.g., if I recall, `torch.stack()` and `tf.stack()` were roughly equivalent to `np.vstack()`, and didn't have anything analogous to `stack`?

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -132,6 +164,11 @@ def _apply_operation(self, operation):
             self._apply_basis_state(operation.parameters[0], wires)
             return
 
+        apply_func = self._ops_map.get(operation.name, None)
+        if apply_func:
+            self._state = apply_func(self._state, axes, inverse=operation.inverse)
+            return
+
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/devices/default_qubit.py:171]: 👍 

An alternate approach is to change them to methods, with a signature like:

```python
def _apply_y(self, state, axes, roll_fn=np.roll, stack_fn=np.stack):
```

e.g., you pass the functions to use for rolling and stacking. But, I don't think this is any cleaner then the current approach (it's probably worse).

File: pennylane/devices/default_qubit.py [hunk 3]
```diff
@@ -132,6 +164,11 @@ def _apply_operation(self, operation):
             self._apply_basis_state(operation.parameters[0], wires)
             return
 
+        apply_func = self._ops_map.get(operation.name, None)
+        if apply_func:
+            self._state = apply_func(self._state, axes, inverse=operation.inverse)
+            return
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:170]: Since these are methods, I would suggest removing the need to pass both `self._state` and return the state:

```suggestion
        apply_func = self._ops_map.get(operation.name, None)
        if apply_func:
            apply_func(axes, inverse=operation.inverse)
            return
```

This is also more consistent with the rest of default.qubit.

Slight nitpick, but I tend to prefer the following pattern for dispatching functions from dictionaries, in being easier to read/parse:

```python
if operation.name in self._ops_map:
    self._ops_map[operation.name](axes, inverse=operation.inverse)
```

There is additional overhead in that the dictionary is being accessed twice (which is fixed with the walrus operator in Python 3.8 😆), but I think this might be negligible?

File: pennylane/devices/default_qubit.py [hunk 4]
```diff
@@ -95,6 +114,18 @@ def __init__(self, wires, *, shots=1000, analytic=True):
         self._state = self._create_basis_state(0)
         self._pre_rotated_state = self._state
 
+        self._ops_map = {
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:117]: I'm thinking maybe `self._ops_apply_fn` or `self._apply_fns` might be more descriptive here? The `map` is a bit redundant, and doesn't really describe what the mapping _is_.

File: pennylane/devices/default_qubit.py [hunk 5]
```diff
@@ -142,6 +179,142 @@ def _apply_operation(self, operation):
...
+        Args:
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        state_x = self._apply_x(state, axes)
+        state_z = self._apply_z(state, axes)
+        return SQRT2INV * (state_x + state_z)
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:232]: nice!

Comment (josh146) [reply pennylane/devices/default_qubit.py:232]: Ah, is the need to return linear combinations of other apply functions the motivation to not having them modify the state in place? 🤔

Comment (josh146) [reply pennylane/devices/default_qubit.py:232]: I wonder if it's possible to replace this with in-place computation, to avoid storing two statevectors in memory...

File: pennylane/devices/default_qubit.py [hunk 6]
```diff
@@ -142,6 +179,142 @@ def _apply_operation(self, operation):
...
+        Returns:
+            array[complex]: output state
+        """
+        sl_0 = _get_slice(0, axes[0], self.num_wires)
+        sl_1 = _get_slice(1, axes[0], self.num_wires)
+
+        if axes[1] > axes[0]:
+            target_axes = [axes[1] - 1]
+        else:
+            target_axes = [axes[1]]
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:257]: This could benefit from a few line comments 😆

File: pennylane/devices/default_qubit.py [hunk 7]
```diff
@@ -142,6 +179,142 @@ def _apply_operation(self, operation):
...
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        all_axes = list(range(len(state.shape)))
+        all_axes[axes[0]] = axes[1]
+        all_axes[axes[1]] = axes[0]
+        return self._transpose(state, all_axes)
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:275]: Awesome!

File: pennylane/devices/default_qubit.py [hunk 8]
```diff
@@ -142,6 +179,142 @@ def _apply_operation(self, operation):
...
+
+        if axes[1] > axes[0]:
+            target_axes = [axes[1] - 1]
+        else:
+            target_axes = [axes[1]]
+
+        state_z = self._apply_z(state[sl_1], axes=target_axes)
+        return self._stack([state[sl_0], state_z], axis=axes[0])
+
+    def _apply_phase(self, state, axes, parameters, inverse=False):
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:299]: Could the `PhaseShift` operation be made to use this method, rather than `apply_diagonal_unitary`?

File: tests/devices/test_default_qubit.py [hunk 9]
```diff
@@ -1733,3 +1734,15 @@ def test_wires_expval(self, wires1, wires2, tol):
...
+def test_get_slice():
+    """Test that the _get_slice function returns the expected slice and allows us to slice
+    correctly into an array."""
+
+    sl = _get_slice(1, 1, 3)
+    array = np.arange(27).reshape((3, 3, 3))
+    target = array[:, 1, :]
+
+    assert sl == (slice(None, None, None), 1, slice(None, None, None))
+    assert np.allclose(array[sl], target)
```
Comment (josh146) [inline tests/devices/test_default_qubit.py:1748]: Nice! Is it worth testing this for additional edge cases? E.g., 1D arrays, arrays with dimension of size 1, e.g., ``[[0, 1]]`

File: tests/devices/test_default_qubit_autograd.py [hunk 10]
```diff
@@ -63,12 +63,15 @@ def test_correct_state(self, tol):
...
+            qml.RZ(np.pi / 4, wires=0)
             return qml.expval(qml.PauliZ(0))
 
         circuit()
         state = dev.state
 
-        expected = np.array([1.0, 0, 1.0, 0]) / np.sqrt(2)
+        amplitude = np.exp(-1j * np.pi / 8) / np.sqrt(2)
+
+        expected = np.array([amplitude, 0, np.conj(amplitude), 0])
```
Comment (josh146) [inline tests/devices/test_default_qubit_autograd.py:74]: What is the motivation behind this change?

**Review by josh146 — COMMENTED** [review]
Thanks Tom, this is awesome!!

> Edge cases where the new approach is slower than the old?

Yeah, I find it a bit strange that some gates seem slower with some interfaces? E.g., CZ is slower on `d.q.tf`, and Hadamard/PauliY seem slower on `d.q.autograd`. Is this something to be concerned over? Since a lot of our demos etc. use smaller circuits, could this increase run time?

> Do we need to test these methods explicitly? We don't for other methods in DefaultQubit.

Yep, this is a huge oversight with default.qubit, we never wrote unit tests!! And because there never were any unit tests, PRs over the years that have modified/improved default.qubit also never added unittests :(

Now that we have the shared device tests, we should ideally remove any duplicated integration tests in `test_default.qubit.py`, and retain only (a) default qubit specific integration tests, and (b) unit tests to test the methods.

> We don't speed up any gates with trainable parameters such as Rot, which are quite common. I don't think there is any fundamental roadblock for doing this though, could be a follow up PR.

Sounds good 👍 The only one that _might_ be worth doing now is `PhaseShift`, since it could potentially just use `apply_phase`. But happy to leave parametrized unitaries to a future PR.

**Comment by trbromley** [comment]
I did a second round of benchmarks, this time on a server rather than my laptop to minimize the influence of other running processes.

image
image
image

For me, this leaves the following gates up for discussion:
- CZ
- Hadamard
- PauliY

Luckily, CZ and PauliY are less commonly used, but Hadamard is pretty common. One thing we can do is prevent the special methods for these gates from occurring in `default.qubit.autograd` and/or `default.qubit.tf`.

For `default.qubit.tf`, I propose to prevent CZ (1355e97bb958ebf3f13845befbd3bf12865d7e0b)
For `default.qubit.autograd` I propose to prevent all three (0d5edb4d00ada06908a35b8cfc39a2d6ef15e242)
For `default.qubit`, let's keep all the methods!

File: pennylane/_qubit_device.py (hunk 1 above)
Comment (josh146) [reply pennylane/_qubit_device.py:88]: Nice!

File: pennylane/devices/default_qubit.py (hunk 5 above)
Comment (josh146) [reply pennylane/devices/default_qubit.py:233]: Sounds good!

File: pennylane/devices/default_qubit_autograd.py [hunk 11]
```diff
@@ -106,6 +106,14 @@ class DefaultQubitAutograd(DefaultQubit):
...
     _conj = staticmethod(np.conj)
     _imag = staticmethod(np.imag)
+    _roll = staticmethod(np.roll)
+    _stack = staticmethod(np.stack)
+
+    def __init__(self, wires, *, shots=1000, analytic=True):
+        super().__init__(wires, shots=shots, analytic=analytic)
+        del self._apply_ops["PauliY"]
+        del self._apply_ops["Hadamard"]
+        del self._apply_ops["CZ"]
```
Comment (josh146) [inline pennylane/devices/default_qubit_autograd.py:116]: 😆 This is to give better benchmarks?

**Review by josh146 — APPROVED** [review]
Looks great! 🎉

File: pennylane/devices/default_qubit_autograd.py (hunk 11 above)
Comment (trbromley) [reply pennylane/devices/default_qubit_autograd.py:116]: Yes :sweat_smile:

Comment (antalszava) [reply pennylane/devices/default_qubit_autograd.py:116]: How so? :smile:

File: pennylane/devices/default_qubit.py [hunk 12]
```diff
@@ -31,8 +31,27 @@
 
 # tolerance for numerical errors
 tolerance = 1e-10
+SQRT2INV = 1 / np.sqrt(2)
+TPHASE = np.exp(1j * np.pi / 4)
 
 
+def _get_slice(index, axis, num_axes):
+    """Allows slicing along an arbitrary axis of an array or tensor.
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:39]: Could be worth adding an example, would also help understanding the exact role of each argument.

Comment (antalszava) [reply pennylane/devices/default_qubit.py:39]: Also, in the quantum sense, is it fair to say that this is a `_get_subsystem` function? If so and didn't miss anything, might be worth a renaming, just because that it might make it more intuitive when reading how the operations are being applied

File: pennylane/devices/default_qubit.py [hunk 13]
```diff
@@ -142,6 +180,148 @@ def _apply_operation(self, operation):
...
+        # If axis[1] is larger than axis[0], then state[sl_1] will have self.num_wires - 1 axes with
+        # the target axes shifted down by one. Otherwise, if axis[1] is less than axis[0] then its
+        # axis number in state[sl_1] remains unchanged. For example: state has axes [0, 1, 2, 3] and
+        # axis[0] = 1 and axis[1] = 3. Then, state[sl_1] has axes [0, 1, 2] so that the target axis
+        # has shifted from 3 to 2. If axis[0] = 2 and axis[1] = 1, then state[sl_1] has axes
+        # [0, 1, 2] but with the target axis remaining unchanged.
+        if axes[1] > axes[0]:
+            target_axes = [axes[1] - 1]
+        else:
+            target_axes = [axes[1]]
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:264]: This part is not entirely clear to me when going for it for some time.

> then state[sl_1] will have self.num_wires - 1 axes with
        # the target axes shifted down by one

Why is it so?

One-piece that is missing is how the first wire is being the control (although the operation does seem to be equivalent). Might be worth explaining perhaps in the main docstring, how using slices and the axes we can end up with the same operation.

File: pennylane/devices/default_qubit.py [hunk 14]
```diff
@@ -142,6 +180,148 @@ def _apply_operation(self, operation):
...
+            state (array[complex]): input state
+            axes (List[int]): target axes to apply transformation
+
+        Returns:
+            array[complex]: output state
+        """
+        sl_0 = _get_slice(0, axes[0], self.num_wires)
+        sl_1 = _get_slice(1, axes[0], self.num_wires)
+
+        # If axis[1] is larger than axis[0], then state[sl_1] will have self.num_wires - 1 axes with
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:255]: Throughout the text
```suggestion
        # If axes[1] is larger than axes[0], then state[sl_1] will have self.num_wires - 1 axes with
```

File: pennylane/devices/default_qubit.py [hunk 15]
```diff
@@ -142,6 +180,148 @@ def _apply_operation(self, operation):
         else:
             self._apply_unitary(matrix, wires)
 
+    def _apply_x(self, state, axes, **kwargs):
+        """Applies a PauliX gate by rolling 1 unit along the axis specified in ``axes``.
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:184]: Might be worth elaborating on why this is equivalent to applying a PauliX

File: pennylane/devices/default_qubit.py (hunk 12 above)
Comment (trbromley) [reply pennylane/devices/default_qubit.py:39]: Thanks, have added an example: 9efdbd5c71399cb913e41f7459ac4a0cf3b7c0f9.

Comment (trbromley) [reply pennylane/devices/default_qubit.py:39]: > is it fair to say that this is a _get_subsystem function?

No, it's slightly different. In a quantum sense, suppose we have a three qubit system with state tensor having shape `(2, 2, 2)`. Then `_get_slice(1, 1, 3)` gives us all of the amplitudes that correspond to having a |1> state in qubit 1. Likewise, `_get_slice(1, 2, 3)` would be all the amplitudes with a |1> state in qubit 2, while `_get_slice(0, 1, 3)` would be all the amplitudes with a |0> state in qubit 1. So really what `_get_slice()` allows you to do is to slice into the state so that your selecting |0> or |1> for a given qubit. This is relevant when, for example, we want to do control or when we want to apply a phase on the |1> state.

Unfortunately, I'm not sure of the best name :laughing:

File: pennylane/devices/default_qubit.py (hunk 15 above)
Comment (trbromley) [reply pennylane/devices/default_qubit.py:197]: Thanks, I tried to add an explanation, but I have to admit these things are hard to explain!

789a724bbd847a160c633ecf77d872c61c2c9c48

File: pennylane/devices/default_qubit.py (hunk 14 above)
Comment (trbromley) [reply pennylane/devices/default_qubit.py]: Thanks, fixed it!

File: pennylane/devices/default_qubit_autograd.py (hunk 11 above)
Comment (trbromley) [reply pennylane/devices/default_qubit_autograd.py:119]: Added: 26e9ab3b4268bae09c2abc3671f2b129ff1ae5c8

File: pennylane/devices/default_qubit.py (hunk 13 above)
Comment (trbromley) [reply pennylane/devices/default_qubit.py]: :thinking: I've tried to explain a bit better, and also add to the main docstring.
7059c5096147436bd933caa1d4410b7956736b32

File: pennylane/devices/default_qubit.py [hunk 16]
```diff
@@ -142,6 +193,157 @@ def _apply_operation(self, operation):
...
+        """
+        sl_0 = _get_slice(0, axes[0], self.num_wires)
+        sl_1 = _get_slice(1, axes[0], self.num_wires)
+
+        # We will be slicing into the state according to state[sl_1], giving us all of the
+        # amplitudes with a |1> for the control qubit. The resulting array has lost an axis
+        # relative to state and we need to be careful about the axis we apply the PauliX rotation
+        # to. If axes[1] is larger than axes[0], then we need to shift the target axis down by
+        # one, otherwise we can leave as-is. For example: a state has [0, 1, 2, 3], control=1,
+        # target=3. Then, state[sl_1] has 3 axes and target=3 now corresponds to the second axis.
```
Comment (antalszava) [inline pennylane/devices/default_qubit.py:282]: :rocket:
